### PR TITLE
Upload a file in terminal

### PR DESCRIPTION
**Pull request by rikasun**
- upload file success

https://user-images.githubusercontent.com/40652126/221294363-0c50fcea-c8e2-438b-8548-fff26fcdef97.mp4

- Upload a file too big failed

https://user-images.githubusercontent.com/40652126/221294267-0180a45f-4e24-4d07-9b15-14fbb9795181.mp4

### CODE REVIEW

File: apps/shell/src/app/modules/prompt/upload/prompt-upload.tsx [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  );
+
+  const onDropRejected = useCallback(() => {
+    console.error('File larger than 1mb');
+  }, []);
+
+  const { getRootProps, getInputProps, isDragActive } = useDropzone({
+    onDrop,
+    onDropRejected,
+    maxSize: 1000000,
```
Comment (rikasun) [inline apps/shell/src/app/modules/prompt/upload/prompt-upload.tsx:24]: tornado buffer size limit 100MB however it shows content length too long if we upload file larger than 1mb, so I limit the size here.

File: libs/redux/src/lib/stores/prompt/prompt.store.ts [hunk 2]
```diff
@@ -281,4 +286,36 @@ export const promptStore: IPromptStore = {
...
   ),
+
+  uploadFile: thunk(
+    async (
+      _actions,
+      { slug, file },
+      { injections: { promptService }, dispatch },
+    ) => {
+      // istanbul ignore next
+      if (!file) {
```
Comment (rikasun) [inline libs/redux/src/lib/stores/prompt/prompt.store.ts:297]: There will be no file if file is rejected by dropzone

**Comment by rikasun** [comment]
We originally use Dropzone.js (https://docs.dropzone.dev/getting-started/setup/imperative), but it doesn't work well with Typescript and redux store. So I used React-dropzone (https://react-dropzone.js.org/) instead, it is quick to connect but I have to set a small area on the terminal just for the upload, because the component will cover our terminal if we set it full screen. 

We probably need to consolidate the top right tool bar in terminal later, but this also explicitly shows our available features.

File: apps/shell/src/app/modules/prompt/upload/prompt-upload.tsx (hunk 1 above)
Comment (rikasun) [reply apps/shell/src/app/modules/prompt/upload/prompt-upload.tsx:24]: still trying to figure out where to increase the size in tornado (https://www.tornadoweb.org/en/stable/_modules/tornado/httputil.html#parse_request_start_line). 1mb seems too small, but it is feature paralled

File: apps/shell/src/app/modules/notification-list/notification-list.tsx [hunk 3]
```diff
@@ -10,9 +10,9 @@ export function NotificationList() {
 
   const printErrors = useMemo(
     () =>
-      notifications.map(({ type, message, id }) => {
+      notifications.map(({ type = NotificationType.Error, message, id }) => {
         const className = clsx(
-          'alert d-block mb-1 border bg-red-600 p-2 text-sm text-white shadow-lg',
```
Comment (rikasun) [inline apps/shell/src/app/modules/notification-list/notification-list.tsx:15]: bg-green at later position should overide bg-red earlier,  not sure why it doesn't, so it is showing success with red background

**Comment by ashblue** [comment]
@rikasun do they have to drag and drop right on top of the "Upload" button? They should be able to drag and drop anywhere on the page. Usually there is an overlay GUI for drag'n drop that appears too and takes over the whole page. There is a file event you can hijack to turn on a drop pad easily https://developer.mozilla.org/en-US/docs/Web/API/HTML_Drag_and_Drop_API/File_drag_and_drop

File: apps/shell/src/app/modules/prompt/upload/prompt-upload.tsx (hunk 1 above)
Comment (ashblue) [reply apps/shell/src/app/modules/prompt/upload/prompt-upload.tsx:24]: Yeah 1mb seems absurdly small for a CLI upload. Might be worth filing a bug and moving to Todo if it's troublesome to tweak the file upload limit.

**Comment by ashblue** [comment]
> We originally use Dropzone.js (https://docs.dropzone.dev/getting-started/setup/imperative), but it doesn't work well with Typescript and redux store. So I used React-dropzone (https://react-dropzone.js.org/) instead, it is quick to connect but I have to set a small area on the terminal just for the upload, because the component will cover our terminal if we set it full screen.
> 
> We probably need to consolidate the top right tool bar in terminal later, but this also explicitly shows our available features.
> 
> CleanShot 2023-02-24 at 13 45 49@2x

I don't think we need React Dropzone. You should be able to reproduce the functionality with two simple event hooks `ondrop` and `ondragover` (see my MDN link above). If you want any pairing with the event hooks I'm happy to hop on a call. 

Snyk Advisor rates the package at 86 which is good. But the library syntax looks more complicated to me than using the vanilla JS event hooks.

**Comment by rikasun** [comment]
Uploading CleanShot 2023-02-27 at 16.05.40.mp4…

@ashblue updated! I accidentally got inspired by the drag and drop section in the network console and styled the dropzone a little bit.

**Comment by ashblue** [comment]
> @ashblue updated! I accidentally got inspired by the drag and drop section in the network console and styled the dropzone a little bit.

That looks really good 👍 . My only suggestion would be to put some easy to read text in the center of the drop zone that says "Drop File Here" or something like that. Beyond that it looks stellar

**Review by ashblue — APPROVED** [review]
LGTM, minor feedback.

File: apps/shell/src/app/modules/notification-list/notification-list.tsx (hunk 3 above)
Comment (rikasun) [reply apps/shell/src/app/modules/notification-list/notification-list.tsx:15]: yes I remember that and was wondering the same.

File: libs/redux/src/lib/stores/prompt/prompt.store.ts [hunk 4]
```diff
@@ -281,4 +286,26 @@ export const promptStore: IPromptStore = {
       }
     },
   ),
+
+  uploadFile: thunk(
+    async (
+      _actions,
```
[the remark this answers is not in the document]
Comment (rikasun) [reply libs/redux/src/lib/stores/prompt/prompt.store.ts:292]: thanks for noticing! but it must be the copilot 😅

[the remark this answers is not in the document]
Comment (ashblue) [reply libs/redux/src/lib/stores/prompt/prompt.store.ts:292]: Copilot or not, definitely a best practice 👍

File: libs/remotes/src/lib/prompt/prompt-service.ts [hunk 5]
```diff
@@ -126,4 +126,22 @@ export const promptService = {
...
+    formData.append('file', file);
+    formData.append('slug', promptSlug);
+    try {
+      const response = await axiosInstance.post(`/api/uploads`, formData);
+      return response.data;
+    } catch (error) {
+      let reason;
+
+      // istanbul ignore next
+      if (error instanceof AxiosError) {
```
[the remark this answers is not in the document]
Comment (rikasun) [reply libs/remotes/src/lib/prompt/prompt-service.ts:140]: Good catch

**Pull request merged**